code/modules/research/server.dm

```diff
@@ -67,95 +53,38 @@
-		current_temp = -1
-		return
-	var/turf/L = get_turf(src)
-	var/datum/gas_mixture/env
-	if(istype(L))
-		env = L.return_air()
-		// This is from the RD server code.  It works well enough but I need to move over the
-		// sspace heater code so we can caculate power used per tick as well and making this both
-		// exothermic and an endothermic component
-		if(env)
-			var/perc = max((get_env_temp() - temp_tolerance_high), 0) * temp_penalty_coefficient / base_mining_income
-
-			env.adjust_heat(heating_power * perc * heat_gen)
-			air_update_turf()
-			src.air_update_turf()
-		else
-			current_temp = env ? env.return_temperature() : -1
-
-/obj/machinery/rnd/server/proc/get_env_temp()
-	// if we are on and ran though one tick
-	if(working && current_temp >= 0)
-		return current_temp
-	else
-		// otherwise we get the temp from the turf
-		var/turf/L = get_turf(src)
-		var/datum/gas_mixture/env
-		if(istype(L))
-			env = L.return_air()
-		return env ? env.return_temperature() : T20C			// env might be null at round start.  This stops runtimes
-
-/obj/machinery/rnd/server/proc/refresh_working()
-	var/current_temp  = get_env_temp()
-
-	// Once we go over the damage temp, the breaker is flipped
-	// Power is still going to the server
-	if(!overheated && current_temp >= temp_tolerance_damage)
-		investigate_log("[src] overheated!", INVESTIGATE_RESEARCH)		// Do we need this?
-		overheated = TRUE
-
-	// If we are over heated, the server will not restart till
-	// eveything is at a safe temp
-	if(overheated && current_temp <= temp_tolerance_low)
-		overheated = FALSE
-
-	// If we are overheateed, start shooting out sparks
-	// don't shoot them if we have no power
-	if(overheated && !(machine_stat & NOPOWER) && prob(40))
-		do_sparks(5, FALSE, src)
-
-	if(overheated || research_disabled || machine_stat & EMPED || machine_stat & NOPOWER)
-		working = FALSE
+/obj/machinery/rnd/server/proc/toggle_disable()
+	set_machine_stat(machine_stat ^ TURNED_OFF)
+
+/obj/machinery/rnd/server/proc/mine()
+	use_power(active_power_usage, power_channel)
+	var/efficiency = get_efficiency()
+	if(efficiency > 0)
+		return list(TECHWEB_POINT_TYPE_GENERIC = max(base_mining_income * efficiency, 0))
 	else
-		working = TRUE
+		return list(TECHWEB_POINT_TYPE_GENERIC = 0)
 
-	update_icon()
+/obj/machinery/rnd/server/proc/get_temperature()
+	return GetComponent(/datum/component/server).temperature
 
-/obj/machinery/rnd/server/emp_act()
-	. = ..()
-	refresh_working()
+/obj/machinery/rnd/server/proc/get_overheat_temperature()
+	return GetComponent(/datum/component/server).overheated_temp
 
-/obj/machinery/rnd/server/emp_reset()
-	..()
-	refresh_working()
+/obj/machinery/rnd/server/proc/get_warning_temperature()
+	return GetComponent(/datum/component/server).warning_temp
 
-/obj/machinery/rnd/server/proc/toggle_disable()
-	research_disabled = !research_disabled
-	refresh_working()
+/obj/machinery/rnd/server/proc/get_efficiency()
+	return GetComponent(/datum/component/server).efficiency
```

Comment:
Byond can't handle this kind of practice, afaik. You need a type holder to access the value.

Reply:
done